File: hecho con python/app/controllers/curso_controller.py
from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session
from typing import List
from app.config.database import get_db
from app.dtos.curso_dto import CursoCreateDTO, CursoUpdateDTO, CursoResponseDTO
from app.repositories.curso_repository import CursoRepository
from app.repositories.estudiante_repository import EstudianteRepository
from app.repositories.profesor_repository import ProfesorRepository
from app.services.curso_service import CursoService

router = APIRouter(prefix="/cursos", tags=["Cursos"])

def get_curso_service(db: Session = Depends(get_db)) -> CursoService:
    curso_repo = CursoRepository(db)
    est_repo = EstudianteRepository(db)
    prof_repo = ProfesorRepository(db)
    return CursoService(curso_repo, est_repo, prof_repo)

@router.get("", response_model=List[CursoResponseDTO], summary="Obtener todos los cursos")
def obtener_todos_los_cursos(service: CursoService = Depends(get_curso_service)):
    return service.obtener_todos_los_cursos()

@router.get("/{id}", response_model=CursoResponseDTO, summary="Obtener curso por ID")
def obtener_curso_por_id(id: int, service: CursoService = Depends(get_curso_service)):
    return service.obtener_curso_por_id(id)

@router.post("", response_model=CursoResponseDTO, status_code=status.HTTP_201_CREATED, summary="Crear curso")
def crear_curso(dto: CursoCreateDTO, service: CursoService = Depends(get_curso_service)):
    return service.crear_curso(dto)

@router.put("/{id}", response_model=CursoResponseDTO, summary="Actualizar curso")
def actualizar_curso(id: int, dto: CursoUpdateDTO, service: CursoService = Depends(get_curso_service)):
    return service.actualizar_curso(id, dto)

File: hecho con java/src/main/java/com/umanizales/notasya/controllers/CursoController.java
package com.umanizales.notasya.controllers;

import com.umanizales.notasya.dtos.CursoDTO;
import com.umanizales.notasya.services.CursoService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/v1/cursos")
@Tag(name = "Cursos", description = "Endpoints para la gestión y calificación de cursos")
@CrossOrigin(origins = "*")
public class CursoController {

    private final CursoService service;

    public CursoController(CursoService service) {
        this.service = service;
    }

    @GetMapping
    @Operation(summary = "Obtener todos los cursos")
    public ResponseEntity<List<CursoDTO.Response>> obtenerTodos() {
        return ResponseEntity.ok(service.obtenerTodosLosCursos());
    }

    @GetMapping("/{id}")
    @Operation(summary = "Obtener curso por ID")
    public ResponseEntity<CursoDTO.Response> obtenerPorId(@PathVariable Long id) {
        return ResponseEntity.ok(service.obtenerCursoPorId(id));
    }

    @PostMapping
    @Operation(summary = "Crear nuevo curso")
    public ResponseEntity<CursoDTO.Response> crear(@Valid @RequestBody CursoDTO.Create dto) {
        return ResponseEntity.status(HttpStatus.CREATED).body(service.crearCurso(dto));
    }

    @PutMapping("/{id}")
    @Operation(summary = "Actualizar curso existente")
    public ResponseEntity<CursoDTO.Response> actualizar(@PathVariable Long id, @Valid @RequestBody CursoDTO.Update dto) {
        return ResponseEntity.ok(service.actualizarCurso(id, dto));
    }
}
